src/psyclone/psyir/transformations/acc_kernels_trans.py

```diff
@@ -99,32 +106,80 @@ def apply(self, node, options=None):
 
         # Create a directive containing the nodes in node_list and insert it.
         directive = ACCKernelsDirective(
             parent=parent, children=[node.detach() for node in node_list],
-            default_present=default_present)
+            default_present=default_present, async_queue=async_queue)
 
         parent.children.insert(start_index, directive)
 
-    def validate(self, nodes, options=None):
+    @staticmethod
+    def check_async_queue(
+                nodes: List[Node],
+                async_queue: Union[bool, int, Reference, None]
+            ):
+        '''
+        Common function to check that all parent data directives have
+        the same async queue.
+
+        :param node: the nodes in the PSyIR to enclose.
+        :param async_queue: The async queue to expect in parents.
+
+        '''
+        if async_queue is None:
+            return
+
+        # check type
+        if not isinstance(async_queue, (int, Reference)):
+            raise TypeError(f"Invalid async_queue value, expect Reference or "
+                            f"integer or None or False, got : {async_queue}")
+
+        # Perform an additional check whether a queue has been used before.
+        # Note this to work only for the current routine.
+        parent = nodes[0].ancestor(ACCAsyncMixin)
+        if parent is not None:
+            if async_queue != parent.async_queue:
+                raise TransformationError(
+                    f"Cannot apply ACCKernelsTrans with asynchronous "
+                    f"queue '{async_queue}' because a parent "
+                    f"directive specifies queue '{parent.async_queue}'")
+
+        parent = nodes[0].ancestor(Routine)
+        if parent:
+            edata = parent.walk(ACCEnterDataDirective)
+            if edata:
+                if async_queue != edata[0].async_queue:
+                    raise TransformationError(
+                        f"Cannot apply ACCKernelsTrans with asynchronous queue"
+                        f" '{async_queue}' because the containing routine "
+                        f"has an ENTER DATA directive specifying queue "
+                        f"'{edata[0].async_queue}'")
+
+    def validate(
+                self,
+                nodes: List[Node],
+                options: Dict[str, Any] = {}
+            ):
         # pylint: disable=signature-differs
         '''
         Check that we can safely enclose the supplied node or list of nodes
         within OpenACC kernels ... end kernels directives.
 
         :param nodes: the proposed PSyIR node or nodes to enclose in the
                       kernels region.
-        :type nodes: (list of) :py:class:`psyclone.psyir.nodes.Node`
         :param options: a dictionary with options for transformations.
-        :type options: Optional[Dict[str, Any]]
-        :param bool options["disable_loop_check"]: whether to disable the
+            "disable_loop_check": boolean whether to disable the
```

Comment:
Same here and below, I am unsure which format is rendered more clearly in the docs.